File: src/main/java/no/novari/configuration/mapping/ValueCollectionMappingMappingService.java
package no.novari.configuration.mapping;

import no.novari.configuration.model.configuration.dtos.CollectionMappingDto;
import no.novari.configuration.model.configuration.dtos.ValueMappingDto;
import no.novari.configuration.model.configuration.entities.collection.ValueCollectionMapping;
import org.springframework.stereotype.Service;

@Service
public class ValueCollectionMappingMappingService {

    private final ValueMappingMappingService valueMappingMappingService;
    private final ValuesFromCollectionMappingMappingService valuesFromCollectionMappingMappingService;

    public ValueCollectionMappingMappingService(
            ValueMappingMappingService valueMappingMappingService,
            ValuesFromCollectionMappingMappingService valuesFromCollectionMappingMappingService
    ) {
        this.valueMappingMappingService = valueMappingMappingService;
        this.valuesFromCollectionMappingMappingService = valuesFromCollectionMappingMappingService;
    }

    public ValueCollectionMapping toEntity(CollectionMappingDto<ValueMappingDto> valueCollectionMappingDto) {
        return ValueCollectionMapping
                .builder()
                .valueMappings(
                        valueCollectionMappingDto
                                .getElementMappings()
                                .stream()
                                .map(valueMappingMappingService::toEntity)
                                .toList()
                )
                .valuesFromCollectionMappings(
                        valueCollectionMappingDto
                                .getFromCollectionMappings()
                                .stream()
                                .map(valuesFromCollectionMappingMappingService::toEntity)
                                .toList()
                )
                .build();
    }

    public CollectionMappingDto<ValueMappingDto> toDto(ValueCollectionMapping valueCollectionMapping) {
        return CollectionMappingDto
                .<ValueMappingDto>builder()
                .elementMappings(
                        valueCollectionMapping
                                .getValueMappings()
                                .stream()
                                .map(valueMappingMappingService::toDto)
                                .toList()
                )
                .fromCollectionMappings(
                        valueCollectionMapping
                                .getValuesFromCollectionMappings()
                                .stream()
                                .map(valuesFromCollectionMappingMappingService::toDto)
                                .toList()
                )
                .build();
    }

}

File: src/main/kotlin/no/novari/flyt/configuration/mapping/ValueCollectionMappingMappingService.kt
package no.novari.flyt.configuration.mapping

import no.novari.flyt.configuration.model.configuration.dtos.CollectionMappingDto
import no.novari.flyt.configuration.model.configuration.dtos.ValueMappingDto
import no.novari.flyt.configuration.model.configuration.entities.collection.ValueCollectionMapping
import org.springframework.stereotype.Service

@Service
class ValueCollectionMappingMappingService(
    private val valueMappingMappingService: ValueMappingMappingService,
    private val valuesFromCollectionMappingMappingService: ValuesFromCollectionMappingMappingService,
) {
    fun toEntity(valueCollectionMappingDto: CollectionMappingDto<ValueMappingDto>): ValueCollectionMapping =
        ValueCollectionMapping
            .builder()
            .valueMappings(valueCollectionMappingDto.elementMappings.map(valueMappingMappingService::toEntity))
            .valuesFromCollectionMappings(
                valueCollectionMappingDto.fromCollectionMappings.map(
                    valuesFromCollectionMappingMappingService::toEntity,
                ),
            ).build()

    fun toDto(valueCollectionMapping: ValueCollectionMapping): CollectionMappingDto<ValueMappingDto> =
        CollectionMappingDto
            .builder<ValueMappingDto>()
            .elementMappings(valueCollectionMapping.valueMappings.map(valueMappingMappingService::toDto))
            .fromCollectionMappings(
                valueCollectionMapping.valuesFromCollectionMappings.map(
                    valuesFromCollectionMappingMappingService::toDto,
                ),
            ).build()
}
